shell/packages/sandstorm-permissions/permissions.js

```diff
@@ -14,333 +14,790 @@
-  // Determines whether the vertex is allowed to open the grain by searching depth first
-  // for a path of active role assignments leading from the grain owner to the vertex.
+function normalize(membranedPermissionSetMap, desiredPermissions) {
+  // `membranedPermissionSetMap` represents the permissions flowing from a grain owner to a vertex,
+  // exactly in the format of the output of `backpropagateVertex()`.
+  //
+  // `desiredPermissions` is the set of permissions that we are currently interested in.
+  //
+  // Returns a set of clauses in a minimal disjunctive normal form, representing the possible
+  // ways that the desired permissions could be achieved. The result is a map from hashed clause
+  // to objects of the form { clause: Clause, tokensUsed: [String] }.
+
+  const result = {};
 
-  var edges = collectEdges(db, vertex);
-  if (edges.grainDoesNotExist || edges.disallowedAnonymousAccess) {
-    return false;
+  function fullCover(chosen) {
+    if (Object.keys(chosen).length == 0) return false;
+
+    const accum = new PermissionSet([]);
+    for (const hashedClause in chosen) {
+      accum.add(membranedPermissionSetMap[hashedClause].permissions);
+    }
+
+    return desiredPermissions.isSubsetOf(accum);
   }
-  if (edges.grainIsPublic) {
-    return true;
+
+  const chooseAll = {};
+  for (const hashedClause in membranedPermissionSetMap) {
+    const pclause = membranedPermissionSetMap[hashedClause];
+    chooseAll[hashedClause] = { clause: pclause.membrane, tokensUsed: pclause.tokensUsed };
   }
-  var edgesByRecipient = edges.edgesByRecipient;
-  var owner = "OwningAccount";
-  if (!edges.terminalEdge) { return false; }
-  if (owner == edges.terminalEdge.sharer) {
-    return true;
+
+  if (!fullCover(chooseAll)) {
+    return {};
   }
 
-  var openerIdentityId = edges.terminalEdge.sharer;
-  var stackedUsers = {};
-  stackedUsers[openerIdentityId] = true;
-  var userStack = [openerIdentityId];
-
-  while (userStack.length > 0) {
-    var recipient = userStack.pop();
-    var edges = edgesByRecipient[recipient];
-    if (edges) {
-      for (var ii = 0; ii < edges.length; ++ii) {
-        var inEdge = edges[ii];
-        var sharer = inEdge.sharer;
-        if (sharer == owner) {
-          return true;
+  const stack = [chooseAll];
+  // Array of Map<hashedClause, { clause: Clause, tokensUsed: [String] }>.
+
+  while (stack.length > 0) {
+    const chosen = stack.pop();
+    let minimal = true;
+    for (const removedHashedClause in chosen) {
+      const newChosen = {};
+      for (const hashedClause in chosen) {
+        if (hashedClause !== removedHashedClause) {
+          newChosen[hashedClause] = chosen[hashedClause];
         }
-        if (!stackedUsers[sharer]) {
-          userStack.push(sharer);
-          stackedUsers[sharer] = true;
+      }
+
+      if (fullCover(newChosen)) {
+        minimal = false;
+        // TODO(perf): keep track of values we've already explored.
```

Comment:
So, if I'm reading this right (which I might not be!), without implementing this TODO, I think this function is O(n!). *With* implementing this TODO, I think the function is still O(2^n) (a set of size n has 2^n subsets). Either way, this is dangerous. I think this function will need to be rethought.